File: src/files/json_serialization/player.cpp
#include "player.h"

Player::Player(QString name, int score) : name(name), score(score) {}

Player::Player(const Player &other) : name(other.name), score(other.score) {}

Player::Player() : name(""), score(0) {}

Player &Player::operator=(const Player &other) {
  name = other.name;
  score = other.score;
  return *this;
}

Player::~Player() {}

QString Player::getName() const { return name; }

int Player::getScore() const { return score; }

void Player::toJson(QJsonObject &root) const {
  root["name"] = name;
  root["score"] = score;
}

void Player::fromJson(const QJsonObject &root) {
  name = root["name"].toString();
  score = root["score"].toInt();
}

QDebug operator<<(QDebug dbg, const Player &player) {
  dbg.nospace() << "Player(" << player.name << ", " << player.score << ")";
  return dbg.space();
}

bool Player::operator==(const Player &other) const {
  return name == other.name && score == other.score;
}

File: src/files/python/json_serialization/player.py
from json_serializable_interface import JsonSerializableInterface

class Player(JsonSerializableInterface):
    def __init__(self, name="", score=0):
        self.name = name
        self.score = score

    def to_json(self, root):
        root["name"] = self.name
        root["score"] = self.score

    def from_json(self, root):
        self.name = root["name"].toString()
        self.score = root["score"].toInt()

    def __eq__(self, other):
        return self.name == other.name and self.score == other.score

    def __str__(self):
        return "Player(" + self.name + ", " + str(self.score) + ")"

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return hash(self.name) + hash(self.score)

    def __lt__(self, other):
        return self.score < other.score

    def __gt__(self, other):
        return self.score > other.score

    def __le__(self, other):
        return self.score <= other.score

    def __ge__(self, other):
        return self.score >= other.score

    def __ne__(self, other):
        return not (self == other)